### PR TITLE
Enable resolving hosts file

### PR DESCRIPTION
**Pull request by FileGo**
This implements functionality desired in #362. It creates a new resolver, which looks up entries in hosts file and resolves them, if found, otherwise query is passed along to next resolver in queue.

In order to avoid any OS compatibility issues, a `hostsFile` configuration entry is required to point out to the hosts file location. If left empty, hosts resolving is disabled.

PS: Sorry I made a bit of a mess with multiple commits, I'm still learning Git :)

### CODE REVIEW

File: resolver/hosts_file_resolver.go [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+	if question.Qtype == dns.TypePTR {
+		response := new(dns.Msg)
+		response.SetReply(request.Req)
+
+		for _, host := range r.hosts {
+			raddr, _ := dns.ReverseAddr(host.IP.String())
+
+			if raddr == question.Name {
+				ptr := new(dns.PTR)
+				ptr.Ptr = dns.Fqdn(host.Hostname)
```
Comment (0xERR0R) [inline resolver/hosts_file_resolver.go:35]: I'm not sure: maybe also aliases should be considered here. Example:

`10.0.0.1                router0 router1 router2`

Should reverse DNS for the IP return only router0 as A record or 3 A records (rourter0, router1 and router2)?

File: resolver/hosts_file_resolver.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+	"os"
+	"strings"
+
+	"github.com/0xERR0R/blocky/model"
+	"github.com/0xERR0R/blocky/util"
+	"github.com/miekg/dns"
+	"github.com/sirupsen/logrus"
+)
+
+const hostsDNSTTL = 3600
```
Comment (0xERR0R) [inline resolver/hosts_file_resolver.go:15]: Please externalize this parameter like "CustomTTL" with defined default value. See https://github.com/0xERR0R/blocky/blob/9259c5cdcdccbda7fd243d33aa564f0153d8c0ef/config/config.go#L313

File: resolver/hosts_file_resolver_test.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+	. "github.com/0xERR0R/blocky/helpertest"
+	. "github.com/0xERR0R/blocky/model"
+	"github.com/miekg/dns"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/stretchr/testify/mock"
+)
+
+var _ = Describe("HostsFileResolver", func() {
```
Comment (0xERR0R) [inline resolver/hosts_file_resolver_test.go:15]: :+1: 100% coverage

**Review by 0xERR0R — CHANGES_REQUESTED** [review]
Thank you for your work! I reviewed your changes and tested a bit, it looks really good! I fixed some lint issues and added some comments/suggestions as review. Feel free to comment.

**Comment by 0xERR0R** [comment]
> 
> PS: Sorry I made a bit of a mess with multiple commits, I'm still learning Git :)

It is ok, I will squash all commits to one on merging this PR

File: resolver/hosts_file_resolver.go (hunk 2 above)
Comment (FileGo) [reply resolver/hosts_file_resolver.go]: Done

File: resolver/hosts_file_resolver.go (hunk 1 above)
Comment (FileGo) [reply resolver/hosts_file_resolver.go:40]: Ah, I've thought you should only get one PTR record, but DNS does allow for more. That's been fixed.

File: resolver/hosts_file_resolver.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+func (r *HostsFileResolver) Resolve(request *model.Request) (*model.Response, error) {
+	logger := withPrefix(request.Log, "hosts_file_resolver")
+
+	if r.HostsFilePath == "" {
+		logger.WithField("resolver", Name(r.next)).Trace("go to next resolver")
+		return r.next.Resolve(request)
+	}
+
+	if !r.parsed {
```
[the remark this answers is not in the document]
Comment (FileGo) [reply resolver/hosts_file_resolver.go]: Makes sense, I was thinking of the same, but didn't know how to implement it. I've look how blocky does it elsewhere and found some good examples.

Refresh period is also user-defined now.

**Comment by 0xERR0R** [comment]
Thanks! It looks good, I added one test case and improved logging a bit. We can merge it :+1:

**Pull request merged**